PEP 800: Solid bases in the type system (#4505)
python/peps: peps/pep-0800.rst

of the solid base mechanism.

Specification
=============

A decorator ``@typing.solid_base`` is added to the type system. It may only be used on nominal classes, including ``NamedTuple``
definitions; it is a type checker error to use the decorator on a function, ``TypedDict`` definition, or ``Protocol`` definition.

We define two properties on (nominal) classes: a class may or may not *be* a solid base, and every class must *have* a valid solid base.

A class is a solid base if it is decorated with ``@typing.solid_base``, or if it contains a non-empty ``__slots__`` definition.
The universal base class, ``object``, is also a solid base.

To determine a class's solid base, we look at all of its base classes to determine a set of candidate solid bases. For each base
that is itself a solid base, the candidate is the base itself; otherwise, it is the base's solid base. If the candidate set contains
a single solid base, that is the class's solid base. If there are multiple candidates, but one of them is a subclass of all other candidates,
that class is the solid base. If no such candidate exists, the class does not have a valid solid base, and therefore cannot exist.

> Type checkers must check for a valid social base when checking class definitions, and emit a diagnostic if they encounter a class

Author: lol, I'm not sure how that slipped in

PEP: 800

> Title: Solid bases in the type system

Reviewer: Nit: PEPs typically use Title casing (first letter of each word capitalized).

decorator, ``@typing.solid_base``, that indicates that a class is a "solid base". Two classes that have distinct, unrelated
solid bases cannot have a common child class.

Motivation
==========

In type checking Python, an important concept is that of reachability. Python type checkers generally
detect when a branch of code can never be reached, and and they warn users about such code. This is useful
because unreachable code unnecessarily complicates the program, and its presence can be an indication of a bug.

For example, in this program::

    def f(x: bool) -> None:
        if isinstance(x, str):
            print("It's both!")

both pyright and mypy (with ``--warn-unreachable``), two popular type checkers, will warn that the body of the
``if`` block is unreachable, because if ``x`` is a ``bool``, it cannot also be a ``str``.

Reachability is complicated in Python by the presence of multiple inheritance. If instead of ``bool`` and ``str``,
we use two user-defined classes, mypy and pyright do not show any warnings::

    class A: pass
    class B: pass

    def f(x: A):
        if isinstance(x, B):
            print("It's both!")

> This is correct, because there a class that inherits from both ``A`` and ``B`` could exist.

Reviewer: "there a class" => "there is a class"

Reviewer 2: I think it would make more sense here to say

> This is correct, because a class that inherits from both ``A`` and ``B`` could exist.

We see a divergence between the type checkers in another case, where we use ``int`` and ``str``::

    def f(x: int):
        if isinstance(x, str):
            print("It's both!")

> On this code, pyright shows no errors but mypy will claim that the branch is unreachable. Mypy is technically correct

Reviewer: Recommend "On this code" => "For this code"

here: CPython does not allow a class to inherit from both ``int`` and ``str``, so the branch is unreachable.
However, the information necessary to determine that these base classes are incompatible is not currently available in

> the type system. Mypy in fact uses a heuristic based on the presence of incompatible methods; this heuristic works

Reviewer: Recommend "Mypy in fact" => "Mypy, in fact, "

reasonably well in practice, especially for built-in types, but it is
`incorrect in general <https://github.com/python/mypy/issues/19377>`__.

The experimental ``ty`` type checker uses a third approach that aligns more closely with the runtime behavior of Python:
it recognizes certain classes as "solid bases" that restrict multiple inheritance. Broadly speaking, every class must

> inherit from a unique solid base, and if there is no unique solid base, the class cannot exist; we'll provide a more

Author: I think the existing text is right because `object` is a solid base. But I'll change it.

Runtime restrictions on multiple inheritance
--------------------------------------------

While Python generally allows multiple inheritance, the runtime imposes various restrictions, as documented in
`CPython PR 136844 <https://github.com/python/cpython/pull/136844/files>`__ (hopefully soon to be merged).
Two sets of restrictions, around a consistent MRO and a consistent metaclass, can already be implemented by
type checkers using information available in the type system. The third restriction, around instance layout,
is the one that requires knowledge of solid bases. Classes that contain a non-empty ``__slots__`` definition
are automatically solid bases, as are many built-in classes implemented in C.

Alternative implementations of Python, such as PyPy, tend to behave similarly to CPython but may differ in details,
such as exactly which standard library classes are solid bases. As the type system does not currently contain any
explicit support for alternative Python implementations, this this PEP recommends that stub libraries such as typeshed
use CPython's behavior to determine when to use the ``@solid_base`` decorator. If future extensions to the type system

> add support for alternative implementations (for example, branching on the value of ``sys.implementation``),

Reviewer 2: ```suggestion
add support for alternative implementations (for example, branching on the value of ``sys.implementation.name``),
```

could possibly also link to `sys.implementation` here?

which may be technically possible but not practically plausible.

::

    @solid_base
    class BaseModel:
        # ... General logic for model classes
        pass

    class Species(BaseModel):
        name: str
        # ... more fields

    def process_species(species: Species):
        if isinstance(species, str):  # oops, forgot `.name`
            pass  # type checker should warn about this branch being unreachable
            # BaseModel and str are solid bases, so a class that inherits from both cannot exist

This is similar in principle to the existing ``@final`` decorator, which also acts to restrict subclassing: in stubs, it
is used to mark classes that programmatically disallow subclassing, but in implementation files, it is often used to
indicate that a class is not intended to be subclassed, without runtime enforcement.

``@solid_base`` on special classes
----------------------------------

The ``@solid_base`` decorator is primarily intended for nominal classes, but the type system contains some other constructs that
syntactically use class definitions, so we have to consider whether the decorator should be allowed on them as well, and if so,

> what it means.

Reviewer 2: ```suggestion
what it would mean.
```

For ``Protocol`` definitions, the most consistent interpretation would be that the only classes that can implement the
protocol would be classes that use nominal inheritance from the protocol, or ``@final`` classes that implement the protocol.
Other classes either have or could potentially have a solid base that is not the protocol. This is convoluted and not useful,
so we disallow ``@solid_base`` on ``Protocol`` definitions.

Similarly, the concept of a "solid base" is not meaningful on ``TypedDict`` definitions, as TypedDicts are purely structural types.

Although they receive some special treatment in the type system, ``NamedTuple`` definitions create real nominal classes that can
have child classes, so it makes sense to allow ``@solid_base`` on them and treat them like regular classes for the purposes

> of the solid base mechanism.

Reviewer 2: I agree that there's no good reason to disallow `@solid_base` on a `NamedTuple` class. It might be worth noting, though, that any `NamedTuple` classes already has a solid base in its MRO, since all `NamedTuple` classes inherit from `tuple`, and `tuple` is a solid base.

Author: Added

that class is the solid base. If no such candidate exists, the class does not have a valid solid base, and therefore cannot exist.

Type checkers must check for a valid solid base when checking class definitions, and emit a diagnostic if they encounter a class
definition that lacks a valid solid base. Type checkers may also use the solid base mechanism to determine whether types are disjoint,
for example when checking whether a type narrowing construct like ``isinstance()`` results in an unreachable branch.

Example::

    from typing import solid_base, assert_never

    @solid_base
    class Solid1:
        pass

    @solid_base
    class Solid2:
        pass

    @solid_base
    class SolidChild(Solid1):
        pass

    class C1:  # solid base is `object`
        pass

    # OK: candidate solid bases are `Solid1` and `object`, and `Solid1` is a subclass of `object`.
    class C2(Solid1, C1):  # solid base is `Solid1`
        pass

    # OK: candidate solid bases are `SolidChild` and `Solid1`, and `SolidChild` is a subclass of `Solid1`.
    class C3(SolidChild, Solid1):  # solid base is `SolidChild`
        pass

>     class C4(Solid1, Solid2):  # error: no single solid base

Reviewer 2: could possibly spell this out a bit more

```suggestion
    # error: candidate solid bases are `Solid1` and `Solid2`,
    # but `Solid1` is not a subclass of `Solid2` and `Solid2` is not a subclass of `Solid1`
    class C4(Solid1, Solid2):
```

Author: Went with a wording closer to what you suggested on the CPython docs PR :)

        if isinstance(obj, Solid2):
            assert_never(obj)  # OK: child class of `Solid1` and `Solid2` cannot exist
        if isinstance(obj, C1):
            reveal_type(obj)  # Shows a non-empty type, e.g. `Solid1 & C1`

Runtime implementation
======================

A new decorator, ``@solid_base``, will be added to the ``typing`` module. Its runtime behavior (consistent with
similar decorators like ``@final``) is to set an attribute ``.__solid_base__ = True`` on the decorated object,
then return its argument::

    def solid_base(cls):
        cls.__solid_base__ = True
        return cls

The ``__solid_base__`` attribute may be used for runtime introspection.

It will be useful to validate whether the ``@solid_base`` decorator should be applied in a stub. While
CPython does not document precisely which classes are solid bases, it is possible to replicate the behavior
of the interpreter using runtime introspection
(`example implementation <https://github.com/JelleZijlstra/pycroscope/blob/0d19236e4eda771175170a6b165b0e9f6a211d19/pycroscope/relations.py#L1469>`__).
Stub validation tools, such as mypy's ``stubtest``, could use this logic to check whether the

> ``@solid_base`` decorator is applied to the correct classes in stubs.

Reviewer 2: Practically speaking, I'm not sure the `__solid_base__` attribute will tell you much other than that the class was explicitly decorated with the `@solid_base` decorator. No builtin class will set the attribute, most external C extensions won't either, and any class that defines non-empty slots will be an implicit solid base even if it doesn't use the decorator. A class might also have a metaclass that prevents setting attributes, so not even all classes explicitly decorated with `@solid_base` will necessarily have this attribute set (https://github.com/python/typing_extensions/pull/634#discussion_r2221910727).

If we want third-party tools to be able to introspect a class's solid base at runtime, a better solution would probably be to add an `inspect.get_solid_base` introspection helper that does something similar to your logic in pycroscope. I don't think it's essential to add that as part of the PEP, though it might be useful.

Reviewer 2: My vague worry is that people will misunderstand this attribute and think that it's a reliable way of determining whether a class is a solid base or not. Unlike `@final`, it feels pretty unlikely that we'll see ever much usage of this decorator in `.py` files, so I'd expect the vast majority of solid bases not to have this attribute. (Though to be clear, I agree that it makes sense to make it possible to apply the decorator to classes in `.py` files.)

Still, you're right that there are lots of `@final` classes that also don't have teh `__final__` attribute. I suppose it's fine to make `solid_base` introspectable whenever it is used in `.py` files, and I suppose the `__solid_base__` attribute no more misleading than the `__final__` attribute set by `@final`

`incorrect in general <https://github.com/python/mypy/issues/19377>`__.

The experimental ``ty`` type checker uses a third approach that aligns more closely with the runtime behavior of Python:
it recognizes certain classes as "solid bases" that restrict multiple inheritance. Broadly speaking, every class must
inherit from at most one unique solid base, and if there is no unique solid base, the class cannot exist; we'll provide a more
precise definition below. However, ty's approach relies on hardcoded knowledge of particular built-in types.

This PEP proposes an extension to the type system that makes it possible to express when multiple inheritance is not
allowed at runtime: an ``@solid_base`` decorator that marks classes as "solid bases".
This gives type checkers a more precise understanding of reachability, and helps in several concrete areas.

Invalid class definitions
-------------------------

The following class definition raises an error at runtime, because ``int`` and ``str`` are distinct solid bases::

    class C(int, str): pass

Without knowledge of solid bases, type checkers are not currently able to detect the reason why this class

> definition is invalid, though they may detect that if this class were to exist, some of its methods would be incompatible.
> (When it sees this class definition, mypy will point at incompatible definitions of ``__add__`` and several other
> methods.)

Reviewer 2: this still doesn't explain the concept of "incompatible definitions" -- if we wanted to, I think it would look something like this. But again, this might be too long outside of a footnote.

```suggestion
definition is invalid, though they may detect that if this class were to exist, some of its methods would
implicitly override methods in its superclass in a way that would be inconsistent with
the Liskov Substitution Principle.
(For example, when mypy sees this class definition, mypy will point at "incompatible definitions" of ``__add__``:
`int` will appear earlier in the MRO, so ``C.__add__`` will be inherited from `int.__add__`;
but this constitutes an implicit -- and Liskov-incompatible -- override of ``str.__add__``
due to the MRO of ``C``.)
```

Author: I think I'll write an appendix discussing this a bit more.

Reviewer 2: Nice, the appendix is great!

additional fields at the end of the struct. CPython knows how to do this with the ``__dict__``
attribute, which is why it is possible to create a child class of ``float`` that adds a ``__dict__``.

However, there is no way to combine a ``float``, which must have a ``double`` in its struct,
with another C type like ``int``, which stores different data at the same spot. Therefore,
a common subclass of ``float`` and ``int`` cannot exist. We say that ``float`` and ``int``
are solid bases.

Classes implemented in C are solid bases if if they have an underlying struct that stores
data at a fixed offset, and that struct is different from the struct of its parent class.
C classes may also store a variable-size array of data (such as the contents of a string);
if this differs from the parent class, the class also becomes a solid base.
CPython's implementation deduces this from the :c:member:`~PyTypeObject.tp_itemsize`
and :c:member:`~PyTypeObject.tp_basicsize` fields of the type object, which are also
accessible from Python code as the undocumented attributes ``__itemsize__`` and ``__basicsize__``
on type objects.

> Similarly, classes implemented in Python are solid bases if they have ``__slots__``, because

Reviewer 2: Worth linking to the datamodel docs for slots here?

Author: ```suggestion
Similarly, classes implemented in Python are solid bases if they have :py:attr:`~object.__slots__`, because
```
